File: Strivers-A2Z-DSA-Sheet/05.Recursion/3.Try Out All Combos/python/05.N queens.py
"""
=============================================================================
  QUESTION: 51. N-Queens (LeetCode)
=============================================================================

  Place n queens on n×n board, no two attack each other.
  Queens attack: same row, column, diagonal.

  APPROACH: Backtracking — column by column.
  Track: row_set, upper diagonal (row-col), lower diagonal (row+col).
  All cells on same ↘ diagonal have same (row-col).
  All cells on same ↗ diagonal have same (row+col).

  DRY RUN n=4:
    Col 0: Q at (1,0) → Col 1: Q at (3,1) → Col 2: Q at (0,2)
    → Col 3: Q at (2,3) → PUSH [".Q..","...Q","Q...","..Q."] ✓

  Time: O(N!), Space: O(N²)
=============================================================================
"""


def solve(col, n, board, ans, row_set, upper_diag, lower_diag):
    if col == n:
        ans.append(["".join(r) for r in board])
        return

    for row in range(n):
        if row not in row_set and (row - col) not in upper_diag and (row + col) not in lower_diag:
            # Place queen
            row_set.add(row)
            upper_diag.add(row - col)
            lower_diag.add(row + col)
            board[row][col] = 'Q'

            solve(col + 1, n, board, ans, row_set, upper_diag, lower_diag)

            # Backtrack
            board[row][col] = '.'
            row_set.remove(row)
            upper_diag.remove(row - col)
            lower_diag.remove(row + col)


def solve_n_queens(n):
    ans = []
    board = [['.' for _ in range(n)] for _ in range(n)]
    solve(0, n, board, ans, set(), set(), set())
    return ans


# Driver Code
if __name__ == "__main__":
    for sol in solve_n_queens(4):
        for row in sol:
            print(row)
        print()

File: Strivers-A2Z-DSA-Sheet/05.Recursion/3.Try Out All Combos/javascript/05.N queens.js
/*
=============================================================================
  QUESTION: 51. N-Queens (LeetCode)
=============================================================================

  Place n queens on n×n board so no two attack each other
  (same row, column, or diagonal).

  Example: n=4 → [[".Q..","...Q","Q...","..Q."],["..Q.","Q...","...Q",".Q.."]]

=============================================================================
  APPROACH: Backtracking — Column by Column — O(N!) Time, O(N²) Space
=============================================================================

  Place queens one column at a time. For each column, try each row.

  O(1) attack check using hash maps:
  - rowMap[row]                → row taken?
  - upperDiag[row - col + n]  → upper-left diagonal taken?
    (all cells on same ↘ diagonal have same row-col value)
  - lowerDiag[row + col]      → lower-left diagonal taken?
    (all cells on same ↗ diagonal have same row+col value)

  Dry Run n=4:
    col 0: row 1 → col 1: row 3 → col 2: row 0 → col 3: row 2 → PUSH ✓
    col 0: row 2 → col 1: row 0 → col 2: row 3 → col 3: row 1 → PUSH ✓

=============================================================================
*/

function solveNQueens(n) {
    let result = [];
    let board = Array.from({ length: n }, () => new Array(n).fill('.'));
    let rowUsed = {}, upperDiag = {}, lowerDiag = {};

    function solve(col) {
        if (col === n) {
            result.push(board.map(r => r.join('')));
            return;
        }

        for (let row = 0; row < n; row++) {
            if (rowUsed[row] || upperDiag[row - col + n] || lowerDiag[row + col])
                continue;

            // Place queen
            board[row][col] = 'Q';
            rowUsed[row] = upperDiag[row - col + n] = lowerDiag[row + col] = true;

            solve(col + 1);

            // Backtrack
            board[row][col] = '.';
            rowUsed[row] = upperDiag[row - col + n] = lowerDiag[row + col] = false;
        }
    }

    solve(0);
    return result;
}

// ==========================================================================
// DRIVER CODE
// ==========================================================================
let solutions = solveNQueens(4);
for (let sol of solutions) {
    console.log(sol);
    console.log("---");
}
// [".Q..","...Q","Q...","..Q."]
// ["..Q.","Q...","...Q",".Q.."]
